### ISSUE TITLE
Memoryleak in Netty Server

### ISSUE THREAD

**Issue by buchgr** [grpc/grpc-java#336]
When pushing the server hard I am getting 

```
SEVERE: LEAK: ByteBuf.release() was not called before it's garbage-collected. See http://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 0
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:250)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:155)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:91)
	io.grpc.transport.netty.NettyWritableBufferAllocator.allocate(NettyWritableBufferAllocator.java:65)
	io.grpc.transport.netty.NettyWritableBufferAllocator.allocate(NettyWritableBufferAllocator.java:48)
	io.grpc.transport.MessageFramer.commitToSink(MessageFramer.java:240)
	io.grpc.transport.MessageFramer.close(MessageFramer.java:221)
	io.grpc.transport.AbstractStream.closeFramer(AbstractStream.java:136)
	io.grpc.transport.AbstractServerStream.close(AbstractServerStream.java:107)
	io.grpc.ServerImpl$ServerCallImpl.close(ServerImpl.java:434)
	io.grpc.stub.ServerCalls$ResponseObserver.onCompleted(ServerCalls.java:195)
	io.grpc.benchmarks.qps.QpsAsyncServer$TestServiceImpl.unaryCall(QpsAsyncServer.java:177)
	grpc.testing.TestServiceGrpc$2.invoke(TestServiceGrpc.java:193)
	grpc.testing.TestServiceGrpc$2.invoke(TestServiceGrpc.java:188)
	io.grpc.stub.ServerCalls$1$1.onHalfClose(ServerCalls.java:93)
	io.grpc.ServerImpl$ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerImpl.java:480)
	io.grpc.ServerImpl$JumpToApplicationThreadServerStreamListener$2.run(ServerImpl.java:370)
	io.grpc.SerializingExecutor$TaskRunner.run(SerializingExecutor.java:152)
	com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:299)
	io.grpc.SerializingExecutor.execute(SerializingExecutor.java:110)
	io.grpc.ServerImpl$JumpToApplicationThreadServerStreamListener.halfClosed(ServerImpl.java:366)
	io.grpc.transport.AbstractServerStream.halfCloseListener(AbstractServerStream.java:243)
	io.grpc.transport.AbstractServerStream.remoteEndClosed(AbstractServerStream.java:200)
	io.grpc.transport.AbstractStream$1.endOfStream(AbstractStream.java:85)
	io.grpc.transport.MessageDeframer.deliver(MessageDeframer.java:245)
	io.grpc.transport.MessageDeframer.request(MessageDeframer.java:142)
	io.grpc.transport.AbstractStream.requestMessagesFromDeframer(AbstractStream.java:226)
	io.grpc.transport.netty.NettyServerStream.access$000(NettyServerStream.java:47)
	io.grpc.transport.netty.NettyServerStream$1.run(NettyServerStream.java:74)
	io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:322)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:356)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:703)
	java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	java.lang.Thread.run(Thread.java:745)
```

and OOM with a 30GB heap after a few minutes. I guess we have a memory leak :-). Working on it.

**Comment by louiscryan** [comment]
I have a fix. It's caused by the allocation of an 'empty' buffer which is not really empty which is never released.

**Comment by nmittler** [comment]
Fixed by #337
